lib/src/code_generator/objc_block.dart

```diff
@@ -115,13 +121,37 @@ class $name extends _ObjCBlockBase {
   static $voidPtr? _cFuncTrampoline;
 
   /// Creates a block from a Dart function.
+  ///
+  /// This block must be invoked by native code running on the same thread as
+  /// the isolate that registered it. Invoking the block on the wrong thread
+  /// will result in a crash.
   $name.fromFunction(${w.className} lib, ${funcType.getDartType(w)} fn) :
       this._(lib.${builtInFunctions.newBlock.name}(
           _dartFuncTrampoline ??= ${w.ffiLibraryPrefix}.Pointer.fromFunction<
               ${trampFuncType.getCType(w)}>($closureTrampoline
                   $exceptionalReturn).cast(), $registerClosure(fn)), lib);
   static $voidPtr? _dartFuncTrampoline;
+
+''');
+
+    // Listener block constructor is only available for void blocks.
+    if (isVoid) {
+      s.write('''
+  /// Creates a listener block from a Dart function.
+  ///
+  /// This is based on FFI's NativeCallable.listener, and has the same
+  /// capabilities and limitations. This block can be invoked from any thread,
+  /// but only supports void functions, and is not run synchronously. See
+  /// NativeCallable.listener for more details.
```

Comment:
This should probably have some documentation about not keeping the isolate alive.

Also, are there use cases in which users would want to keep the isolate alive? Should we surface a bool on the API?